### PR TITLE
feat: add Argo CD Application

### PR DESCRIPTION
**Pull request by maximilianbraun**
### Description of your changes

* added support for Argo CD Applications
* added conversion of structs via `goverter` (used within Crossplane as well)
* added `mockgen` to be executed with `go generate ./...`
* added 2 make targets for simpler local dev + little shell script
* updated k8s clients, i had some issues with building
* updated `golangci-lint` version, which contained a contained a couple fixes regarding resource usage

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Unit tests have been added to the controllers & manual integration tests. 

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by janwillies** [comment]
Thanks a lot @maximilianbraun for this massive contribution! We'll take a look soon

File: pkg/controller/applications/controller.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+func generateCreateApplicationRequest(p *v1alpha1.ApplicationParameters, name string) *application.ApplicationCreateRequest { // nolint:gocyclo
+	converter := v1alpha1.ConverterImpl{}
+	spec := converter.ToArgoApplicationSpec(p)
+
+	app := &argocdv1alpha1.Application{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:213]: `Name` should always be `meta.GetExternalName(cr)`.

If the external name is not set but needed, the external name should be set by an initializer, like so: https://github.com/crossplane-contrib/provider-aws/blob/master/pkg/controller/docdb/dbcluster/setup.go#L76

File: pkg/controller/applications/controller.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	o := &application.ApplicationUpdateRequest{
+		Application: app,
+	}
+	return o
+}
+
+func isApplicationUpToDate(cr *v1alpha1.ApplicationParameters, remote *argocdv1alpha1.Application) bool { // nolint:gocyclo
+	converter := v1alpha1.ConverterImpl{}
+	cluster := converter.ToArgoApplicationSpec(cr)
+	return reflect.DeepEqual(*cluster, remote.Spec)
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:247]: Best use `cmp.Equal` as it is also used elsewhere.

```suggestion
	return cmp.Equal(*cluster, remote.Spec)
```

File: apis/applications/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	// Sources is a reference to the location of the application's manifests or chart
+	Sources ApplicationSources `json:"sources,omitempty" protobuf:"bytes,8,opt,name=sources"`
+}
+
+// ResourceIgnoreDifferences contains resource filter and list of json paths which should be ignored during comparison with live state.
+type ResourceIgnoreDifferences struct {
+	Group             string   `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+	Kind              string   `json:"kind" protobuf:"bytes,2,opt,name=kind"`
+	Name              string   `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+	Namespace         string   `json:"namespace,omitempty" protobuf:"bytes,4,opt,name=namespace"`
```
Comment (MisterMX) [inline apis/applications/v1alpha1/types.go:56]: Optional fields should be a pointer.

**Review by MisterMX — CHANGES_REQUESTED** [review]

**Comment by maximilianbraun** [comment]
> Thanks for your contribution @maximilianbraun! Can you take a lot a the remarks I have added?

Sure thing. Let me have a look :)

File: Makefile [hunk 4]
```diff
@@ -9,6 +9,7 @@ PLATFORMS ?= linux_amd64 linux_arm64
 # kind-related versions
 KIND_VERSION ?= v0.14.0
 KIND_NODE_IMAGE_TAG ?= v1.24.0
+GOLANGCILINT_VERSION ?= 1.53.1
```
[the remark this answers is not in the document]
Comment (maximilianbraun) [reply Makefile:12]: Oh yeah mate, I had a specific problem with memory consumption on my box which was fixed with raising the version. Let me see how the linter behaves after updating of the `build` tooling

File: pkg/controller/applications/controller.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1alpha1.Application)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotApplication)
+	}
+
+	createRequest := generateCreateApplicationRequest(&cr.Spec.ForProvider, cr.Name)
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:146]:
```suggestion
	createRequest := generateCreateApplicationRequest(cr)
```

File: pkg/controller/applications/controller.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotApplication)
+	}
+
+	var name = meta.GetExternalName(cr)
+
+	if name == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:104]: Can be simplified

```suggestion
		return managed.ExternalObservation{}, nil
```

File: pkg/controller/applications/controller.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		},
+		Spec: *spec,
+	}
+
+	repoCreateRequest := &application.ApplicationCreateRequest{
+		Application: app,
+	}
+
+	return repoCreateRequest
+}
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:222]: Make it simpler by passing `cr` and get the name automatically.

```suggestion
func generateCreateApplicationRequest(p *v1alpha1.Application) *application.ApplicationCreateRequest {
	converter := v1alpha1.ConverterImpl{}
	spec := converter.ToArgoApplicationSpec(&p.Spec.ForProvider)

	app := &argocdv1alpha1.Application{
		TypeMeta: metav1.TypeMeta{},
		ObjectMeta: metav1.ObjectMeta{
			Name: meta.GetExternalName(cr),
		},
		Spec: *spec,
	}

	repoCreateRequest := &application.ApplicationCreateRequest{
		Application: app,
	}

	return repoCreateRequest
}
```

File: pkg/controller/applications/controller.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+			Name: name,
+		},
+		Spec: *spec,
+	}
+
+	o := &application.ApplicationUpdateRequest{
+		Application: app,
+	}
+	return o
+}
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:241]: Make it simpler by passing `cr` and get the name automatically:

```suggestion
func generateUpdateRepositoryOptions(cr *v1alpha1.Application) *application.ApplicationUpdateRequest {
	converter := v1alpha1.ConverterImpl{}

	spec := converter.ToArgoApplicationSpec(&cr.Spec.ForProvider)

	app := &argocdv1alpha1.Application{
		TypeMeta: metav1.TypeMeta{},
		ObjectMeta: metav1.ObjectMeta{
			Name: meta.GetExternalName(cr),
		},
		Spec: *spec,
	}

	o := &application.ApplicationUpdateRequest{
		Application: app,
	}
	return o
}
```

File: pkg/controller/applications/controller.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+	return managed.ExternalCreation{}, errors.Wrap(nil, errKubeUpdateFailed)
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1alpha1.Application)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotApplication)
+	}
+	name := meta.GetExternalName(cr)
+	updateRequest := generateUpdateRepositoryOptions(&cr.Spec.ForProvider, name)
```
Comment (MisterMX) [inline pkg/controller/applications/controller.go:162]:
```suggestion
	updateRequest := generateUpdateRepositoryOptions(cr)
```

**Review by MisterMX — CHANGES_REQUESTED** [review]
@maximilianbraun looking good already. I marked some last things that need to be fixed before merging.

**Review by MisterMX — APPROVED** [review]
LGTM. Thanks a lot @maximilianbraun!

**Pull request merged**